update embedsub.py

bjkuhn opened the pull request:
The previous code created a new fits file and only appended the sci, err, and dq data and header extensions. This was causing problems down the line with astrodrizzle because the D2MIARR extensions were no longer a part of the file (which is a requirement for astrodrizzle). The code now makes a copy of the input file so that all the extensions get carried over to the embedded output file.

dsom-stsci commented:
And we checked this update is in line with what the original embed_subs.py routine did

bjkuhn commented:
@cshanahan1 do you know who I can work with to get this approved and merged?

cshanahan1 commented:
@bjkuhn i can review and merge this one

bjkuhn commented:
@cshanahan1 do you know approximately when you'll be able to approve and merge this?  I ask because once it's merged we'll be able to publish one of our HSTaXe tutorial notebooks

cshanahan1 commented on wfc3tools/embedsub.py:
```diff
@@ -101,18 +101,23 @@ def embedsub(files):
-            hdulist.append(fits.ImageHDU(dq, header=flt[5].header, name='TIME'))
-
-        hdulist.writeto(full, clobber=False)
+        with fits.open(full,mode='update') as hdu:
```
comma after full

cshanahan1 commented on wfc3tools/embedsub.py:
```diff
@@ -101,18 +101,23 @@ def embedsub(files):
         flt[0].header['SUBARRAY'] = False
 
+        # Make a duplicate of the input file and name it variable full
+        os.system(f"cp {filename} {full}")
```
i'm not sure you have to do this - i think if you open the input file in read-only mode, modify its hdu list, you can use the hdu.write_to method to write that out to a new output file

cshanahan1 reviewed (commented):
this is my first time looking at this code and I notice some possible issues with it unrelated to the changes you've made that should be addressed in the future. one of them is that this is assuming a pre-determined number of extensions - 4 for UVIS which makes sense but only 6 for IR - wouldn't additional reads past 6 be ignored if this code is run in its current state? 

other than that, i think this fix is fine for now i only left a couple of comments.

bjkuhn commented:
@cshanahan1 thank you for the comments. I addressed all of them, including the problem you found regarding the pre-determined number of extensions. Let me know what you think and if any more changes are needed 

two main changes from commit [053fa56](https://github.com/spacetelescope/wfc3tools/pull/70/commits/053fa568b3e3eb1aeec0c92cf4a4cfc47761255b):
1.  rather than have a hard-coded range for which fits extensions should be edited I loop through all possible extensions and check if the keywords we want to change already exist in the i'th extension. If the keyword exists it is modified, if it does not exist the loop passes to the next extension. 
2. instead of copying the original file and modifying the copied file, I open the original file in read-only and make the changes needed before writting the hdulist to the new file name

cshanahan1 commented on wfc3tools/embedsub.py:
```diff
@@ -91,28 +91,43 @@ def embedsub(files):
-            flt[i].header['ltv2'] = 0.0
+        # loop through each possible fits extension and modify keywords if they exist
+        for i in range(1, len(flt)):
+            try:
```
instead of a try except here, you could just check 'if crpix1 in flt[i].header.keys'

cshanahan1 commented on wfc3tools/embedsub.py:
```diff
@@ -91,28 +91,43 @@ def embedsub(files):
-        hdulist.append(fits.ImageHDU(err, header=flt[2].header, name='ERR'))
-        hdulist.append(fits.ImageHDU(dq, header=flt[3].header, name='DQ'))
-
+        hdu_list = fits.open(filename)
```
since you're now iterating above (in case there are more than 5 science extensions), i would do the same here.

cshanahan1 reviewed (commented):
made a few more comments. once those are addressed, i think we can merge this. since there are no tests, ill assume you've tested it yourself. we should write some in the future for this module

bjkuhn commented:
@cshanahan1 thank you again for the comments on the last commit.  I have updated the code again ([4a6a301](https://github.com/spacetelescope/wfc3tools/pull/70/commits/4a6a301ea960006d5f357623c963cbfd8d89df29)) to incorporate the comments you made above.

two main updates: 
1.  removed unnecessary try block for setting the header keywords `crpix1`, `crpix2`, `ltv1`, and `ltv2`. 
      note:  `if 'crpix1' in flt[i].header.keys` does not work but `if 'crpix1' in flt[i].header` does
2.  now the code iterates through each extension and saves all the headers in case there are more than 5 extensions

testing: 
yes, each time I make edits I have been testing embedsub on IR file: `if1z02akq` and UVIS file: `iels01aaq`

cshanahan1 commented:
i think i was misunderstanding something - since this is for IR flts only, not IMAs, then there will only be one IR science extension (and one dq and err extension as well). sorry it has been a while since i looked at WFC3 data, i forgot that the IR flts were like this. i think we can revert some of the commits then, @bjkuhn i can do this since i asked you to make those changes in the first place

cshanahan1 commented on wfc3tools/embedsub.py:
```diff
@@ -91,28 +91,37 @@ def embedsub(files):
-            hdulist.append(fits.ImageHDU(dq, header=flt[5].header, name='TIME'))
-
-        hdulist.writeto(full, clobber=False)
+        hdu_list = fits.open(filename)
```
i think since 'flt' represents the input file opened in read-only mode already, this line can be omitted. flt[i].header will be the correct header already (no need to do hdu_list[i].header = flt[i].header), and the arrays can be assigned directly to flt[1], flt[2] etc, and then it can be written out. i'm going to change this

cshanahan1 commented:
@bjkuhn can you test out the latest commit i made?

bjkuhn commented:
@cshanahan1 thank you for the commit. I tested the latest changes and everything works as expected

cshanahan1 commented:
ok great, im gonna merge this then!
